examples/transactions_example/processor.go

```diff
@@ -0,0 +1,354 @@
+// all using the transactional API.
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"sync"
+	"time"
+)
+
+// The processor's consumer group id.
+var processorGroupID = "go-transactions-example-processor"
+
+// intersectionState
+type intersectionState struct {
+	name        string            // Name of intersection
+	partition   int32             // Input partition
+	lightState  map[string]string // Current light states, indexed by road
+	carsWaiting map[string]int    // Numbers of cars waiting, indexed by road
+	carCnt      int               // Total number of cars waiting
+	currGreen   string            // Currently green road
+	lastChange  time.Time         // Time of last light change
+}
+
+// intersectionStates maintains state per intersection
+var intersectionStates map[string]*intersectionState
+
+// producers map assigned consumer input partitions to Producer instances.
+var producers map[int32]*kafka.Producer
+
+// consumer is the processor consumer
+var processorConsumer *kafka.Consumer
+
+// lightStateMsg is the state representation of a traffic light and ingress road
+// as sent on the output topic enveloped in intersectionStateMsg.
+type lightStateMsg struct {
+	Name        string // Name of intersection
+	Road        string // Name of ingress road, this in combination with Name identifies a single light
+	State       string // red, green
+	CarsWaiting int    // Number of cars waiting at this ingress
+}
+
+// intersectionStateMsg is the state representation of an intersection
+// with its traffic lights as sent on the output topic as JSON.
+type intersectionStateMsg struct {
+	Name   string                   // Name of intersection
+	Lights map[string]lightStateMsg // Per traffic light state
+}
+
+func (m lightStateMsg) String() string {
+	return fmt.Sprintf("lightState{%s: %s is %s, %d cars waiting}",
+		m.Name, m.Road, m.State, m.CarsWaiting)
+}
+
+func getIntersectionState(name string, partition int32) *intersectionState {
+	istate, found := intersectionStates[name]
+	if found {
+		return istate
+	}
+
+	istate = &intersectionState{
+		name:      name,
+		partition: partition,
+	}
+
+	istate.lightState = make(map[string]string)
+	istate.carsWaiting = make(map[string]int)
+
+	for _, light := range Roads {
+		istate.lightState[light] = "red"
+		istate.carsWaiting[light] = 0
+	}
+
+	intersectionStates[name] = istate
+
+	return istate
+}
+
+// getNextRoad returns the next road in a cyclical fashion.
+func getNextRoad(currRoad string) string {
+	for i, name := range Roads {
+		if currRoad == name {
+			return Roads[(i+1)%len(Roads)]
+		}
+	}
+
+	return Roads[0]
+}
+
+// processIngressCarMessage processes an input message of a car coming
+// into an intersection.
+func processIngressCarMessage(msg *kafka.Message) {
+	if len(producers) == 0 {
+		// Outdated message from last rebalance;
+		// no producers are available.
+		return
+	}
+
+	if msg.Key == nil || msg.Value == nil {
```

Comment:
why check for this? seems a bit arbitrary

Reply:
To avoid a null dereference, we don't control what's coming on the input topic.